libindy/src/services/metrics/mod.rs

```diff
@@ -0,0 +1,113 @@
+const QUEUED_COMMANDS_COUNT: &str = "queued_commands_count";
+const QUEUED_COMMANDS_DURATION_MS: &str = "queued_commands_duration_ms";
+const EXECUTED_COMMANDS_COUNT: &str = "executed_commands_count";
+const EXECUTED_COMMANDS_DURATION_MS: &str = "executed_commands_duration_ms";
+
+const COMMANDS_COUNT: usize = MetricsService::commands_count();
+
+pub struct MetricsService {
+    queued_commands_count: RefCell<[u128; COMMANDS_COUNT]>,
+    queued_commands_duration_ms: RefCell<[u128; COMMANDS_COUNT]>,
+
+    executed_commands_count: RefCell<[u128; COMMANDS_COUNT]>,
+    executed_commands_duration_ms: RefCell<[u128; COMMANDS_COUNT]>,
+}
+
+impl MetricsService {
+    pub fn new() -> Self {
+        MetricsService {
+            queued_commands_count: RefCell::new([u128::MIN; COMMANDS_COUNT]),
+            queued_commands_duration_ms: RefCell::new([u128::MIN; COMMANDS_COUNT]),
+
+            executed_commands_count: RefCell::new([u128::MIN; COMMANDS_COUNT]),
+            executed_commands_duration_ms: RefCell::new([u128::MIN; COMMANDS_COUNT]),
+        }
+    }
+
+    pub fn cmd_left_queue(&self, command_index: CommandIndex, duration: u128) {
+        self.queued_commands_count.borrow_mut()[command_index as usize] += 1;
+        self.queued_commands_duration_ms.borrow_mut()[command_index as usize] += duration;
+    }
+
+    pub fn cmd_executed(&self, command_index: CommandIndex, duration: u128) {
+        self.executed_commands_count.borrow_mut()[command_index as usize] += 1;
+        self.executed_commands_duration_ms.borrow_mut()[command_index as usize] += duration;
+    }
+    pub fn cmd_name(index: usize) -> String {
+        CommandIndex::from(index).to_string().to_case(Case::Snake)
+    }
+    const fn commands_count() -> usize {
+        CommandIndex::VARIANT_COUNT
+    }
+
+    pub fn append_command_metrics(&self, metrics_map: &mut Map<String, Value>) {
+        for index in (0..MetricsService::commands_count()).rev() {
+            let cmd_name = MetricsService::cmd_name(index);
+            metrics_map.insert(format!("{}_{}", cmd_name.as_str(), EXECUTED_COMMANDS_COUNT).as_str().to_string(),
+                               Value::from(self.executed_commands_count.borrow()[index] as usize));
+            metrics_map.insert(format!("{}_{}", cmd_name.as_str(), EXECUTED_COMMANDS_DURATION_MS).as_str().to_string(),
+                               Value::from(self.executed_commands_duration_ms.borrow()[index] as usize));
+            metrics_map.insert(format!("{}_{}", cmd_name.as_str(), QUEUED_COMMANDS_COUNT).as_str().to_string(),
+                               Value::from(self.queued_commands_count.borrow()[index] as usize));
+            metrics_map.insert(format!("{}_{}", cmd_name.as_str(), QUEUED_COMMANDS_DURATION_MS).as_str().to_string(),
+                               Value::from(self.queued_commands_duration_ms.borrow()[index] as usize));
+        }
+    }
+}
+
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_cmd_left_queue() {
```

Comment:
What about naming this `test_cmd_left_queue_increments_relevant_queued_counters`?

Reply:
Done